libdevcore/LevelDB.cpp

```diff
@@ -0,0 +1,169 @@
+/*
+	This file is part of cpp-ethereum.
+
+	cpp-ethereum is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "LevelDB.h"
+#include "Assertions.h"
+
+namespace dev
+{
+namespace db
+{
+namespace
+{
+
+inline leveldb::Slice toLDBSlice(Slice _slice)
+{
+    return leveldb::Slice(_slice.data(), _slice.size());
+}
+
+class LevelDBWriteBatch: public WriteBatchFace
+{
+public:
+	explicit LevelDBWriteBatch(leveldb::WriteOptions _writeOptions): m_writeOptions(std::move(_writeOptions))
+	{
+	}
+
+	void insert(Slice _key, Slice _value) override;
+	void kill(Slice _key) override;
+
+	const leveldb::WriteBatch& writeBatch() const { return m_writeBatch; }
+	leveldb::WriteBatch& writeBatch() { return m_writeBatch; }
+
+private:
+	const leveldb::WriteOptions m_writeOptions;
+	leveldb::WriteBatch m_writeBatch;
+};
+
+void LevelDBWriteBatch::insert(Slice _key, Slice _value)
+{
+	m_writeBatch.Put(toLDBSlice(_key), toLDBSlice(_value));
+}
+
+void LevelDBWriteBatch::kill(Slice _key)
+{
+	m_writeBatch.Delete(toLDBSlice(_key));
+}
+
+}
+
+leveldb::ReadOptions LevelDB::defaultReadOptions()
+{
+	leveldb::ReadOptions readOptions;
+	readOptions.verify_checksums = true;
+	return readOptions;
+}
+
+leveldb::WriteOptions LevelDB::defaultWriteOptions()
+{
+	return leveldb::WriteOptions();
+}
+
+leveldb::Options LevelDB::defaultDBOptions()
+{
+	leveldb::Options options;
+	options.create_if_missing = true;
+	options.max_open_files = 256;
+	return options;
+}
+
+LevelDB::LevelDB(
+	std::string const& _path,
+	leveldb::ReadOptions _readOptions,
+	leveldb::WriteOptions _writeOptions,
+	leveldb::Options _dbOptions
+): m_db(nullptr), m_readOptions(std::move(_readOptions)), m_writeOptions(std::move(_writeOptions))
+{
+	auto db = static_cast<leveldb::DB*>(nullptr);
+	const auto status = leveldb::DB::Open(_dbOptions, _path, &db);
+	if (!status.ok()) {
+		BOOST_THROW_EXCEPTION(FailedToOpenDB(status.ToString()));
+	}
+	if (!db) {
+		BOOST_THROW_EXCEPTION(FailedToOpenDB("null database pointer"));
+	}
+	m_db.reset(db);
+}
+
+std::string LevelDB::lookup(Slice _key) const
+{
+	const leveldb::Slice key(_key.data(), _key.size());
```

Comment:
`const` should be after type name everywhere

Reply:
Right. I wish clang-format could do this transformation.

Reply:
`clang-format` can't do it at the moment. But you should format the code with `clang-format` anyway.

About the placement of the `const` itself, I suggest we rather follow C++ Core Guidelines: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rl-const.

Reply:
Yeah, but again, I wouldn't like to have two styles mixed in the codebase